sound/soc/sof/intel/telemetry.c

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+//
+
+#include <sound/sof/ipc4/header.h>
+#include <sound/sof/xtensa.h>
+#include "../ipc4-priv.h"
+#include "../sof-priv.h"
+#include "hda.h"
+#include "telemetry.h"
+
+void sof_ipc4_intel_dump_telemetry_state(struct snd_sof_dev *sdev, u32 flags)
+{
+	struct telemetry_slot_data *telemetry_data;
+	struct sof_ipc_dsp_oops_xtensa *xoops;
+	u32 slot_offset;
+	char *level;
+
+	slot_offset = sof_ipc4_find_debug_slot_offset_by_type(sdev, SOF_IPC4_DEBUG_SLOT_TELEMETRY);
+	if (!slot_offset)
+		return;
+
+	telemetry_data = kmalloc(sizeof(*telemetry_data), GFP_KERNEL);
+	if (!telemetry_data)
+		return;
+
+	sof_mailbox_read(sdev, slot_offset, telemetry_data, sizeof(*telemetry_data));
+	if (telemetry_data->separator != XTENSA_CORE_DUMP_SEPARATOR) {
+		dev_err(sdev->dev, "Core dump information is not available due to invalid separator %#x\n",
+			telemetry_data->separator);
+		goto free;
+	}
+
+	if (telemetry_data->block.soc != XTENSA_SOC_INTEL_ADSP) {
+		dev_err(sdev->dev, "Core dump information is not available due to invalid SOC %d\n",
+			telemetry_data->block.soc);
+		goto free;
+	}
+
+	if (telemetry_data->hdr.id[0] != COREDUMP_HDR_ID0 ||
+	    telemetry_data->hdr.id[1] != COREDUMP_HDR_ID1 ||
+	    telemetry_data->arch_hdr.id != COREDUMP_ARCH_HDR_ID) {
+		dev_err(sdev->dev, "Core dump information is not available due to invalid coredump header %c%c, arch hdr %c\n",
+			telemetry_data->hdr.id[0], telemetry_data->hdr.id[1],
+			telemetry_data->arch_hdr.id);
+		goto free;
+	}
+
+	switch (telemetry_data->block.toolchain) {
+	case XTENSA_TOOL_CHAIN_ZEPHYR:
+		dev_info(sdev->dev, "FW is built with zephyr toolchain\n");
+		break;
+	case XTENSA_TOOL_CHAIN_XCC:
+		dev_info(sdev->dev, "FW is built with XCC toolchain\n");
+		break;
+	default:
+		dev_warn(sdev->dev, "Unknown toolchain is used\n");
+		break;
+	}
+
+	xoops = kzalloc(flex_array_size(xoops, ar, XTENSA_CORE_AR_REGS_COUNT), GFP_KERNEL);
```

Comment:
this seems wrong - it doesn't include the size of the actual structure, only the flexible array, and causes compilation to fail for me as:
```
/usr/local/src/project/51/src/linux/sound/soc/sof/intel/telemetry.c: In function ‘sof_ipc4_intel_dump_telemetry_state’:
/usr/local/src/project/51/src/linux/sound/soc/sof/intel/telemetry.c:79:2: error: ‘memcpy’ offset 124 is out of the bounds [0, 64] [-Werror=array-bounds]
   79 |  memcpy(xoops->ar, telemetry_data->block.ar,
      |  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |         XTENSA_CORE_AR_REGS_COUNT * sizeof(int));
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```
You need
```
diff --git a/sound/soc/sof/intel/telemetry.c b/sound/soc/sof/intel/telemetry.c
index fa0f51000b80..a596bc3a4528 100644
--- a/sound/soc/sof/intel/telemetry.c
+++ b/sound/soc/sof/intel/telemetry.c
@@ -63,7 +63,8 @@ void sof_ipc4_intel_dump_telemetry_state(struct snd_sof_dev *sdev, u32 flags)
 		break;
 	}
 
-	xoops = kzalloc(flex_array_size(xoops, ar, XTENSA_CORE_AR_REGS_COUNT), GFP_KERNEL);
+	xoops = kzalloc(sizeof(*xoops) + flex_array_size(xoops, ar, XTENSA_CORE_AR_REGS_COUNT),
+			GFP_KERNEL);
 	if (!xoops)
 		goto free;
 
```

Reply:
Right, it is `struct_size(xoops, ar, XTENSA_CORE_AR_REGS_COUNT)`